File: Trie/200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        visited = set()


        no_of_islands = 0

        def dfs(r,c):
            visited.add((r,c))


            dir = [(-1,0), (1, 0),(0,1),(0,-1)]

            for i, j in dir:
                new_r = i + r
                new_c = j + c

                if new_r in range(ROWS) and new_c in range(COLS) and (new_r, new_c) not in visited and grid[new_r][new_c] == '1':
                    dfs(new_r, new_c)


            

        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == '1' and (i,j) not in visited:
                    dfs(i,j)
                    no_of_islands +=1
        return no_of_islands

    

File: Trie/200-number-of-islands/number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const ROWS = grid.length
    const COLS = grid[0].length
    let visited = new Set()
    let islands = 0

    const dfs = (r, c) => {
        visited.add(`${r},${c}`)
        let dir = [[-1,0],[1,0],[0,-1],[0,1]]
        for( let [dx, dy] of dir) {
            const new_r = r + dx
            const new_c = c + dy

            if ( new_r >= 0 && new_r <= ROWS-1 && new_c >= 0 && new_c <= COLS-1 && !visited.has(`${new_r},${new_c}`) && grid[new_r][new_c] === '1') {
                dfs(new_r, new_c)
            }
        }
            
    
    }
    for(let i=0; i<ROWS; i++ ) {
        for (let j=0;j<COLS; j++) {
            if (!visited.has(`${i},${j}`) && grid[i][j] === "1") {
                dfs(i,j)
                islands+=1
            }
        }
    }
    return islands


};